File: app/backend/gap_analysis_engine.py
"""
gap_analysis_engine.py — Python port of GapAnalysisEngine (gap-analysis-engine.js)
Faithful 1-to-1 port: same scoring rules, same thresholds, same output schema.
"""
from indicators import compute_adx, compute_rsi, compute_volume_ratio
from constants import (GAP_MIN_PCT, GAP_TIER2_PCT, GAP_TIER3_PCT, GAP_TIER4_PCT,
                       GAP_W_NORMAL, GAP_W_STRONG, GAP_W_FADE)


class GapAnalysisEngine:

    def __init__(self):
        # { symbol: { fillRate, totalGaps, filledGaps } }
        self.gap_fill_database = {}
        self.reversal_detector  = {}
        self.initialization_flag = False

    # ── Gap Fill History ──────────────────────────────────────────────────────

    def initialize_gap_fill_history(self, symbol, ohlcv):
        if not symbol or symbol in self.gap_fill_database:
            return
        self.build_gap_fill_history(symbol, ohlcv)
        self.initialization_flag = True

    def build_gap_fill_history(self, symbol, ohlcv):
        if not ohlcv or len(ohlcv) < 10:
            return
        total_gaps = filled_gaps = 0
        for i in range(1, len(ohlcv)):
            prev_close = ohlcv[i-1]['close']
            open_price = ohlcv[i]['open']
            gap_pct = abs((open_price - prev_close) / prev_close * 100)
            if gap_pct < 0.5:
                continue
            total_gaps += 1
            is_gap_up = open_price > prev_close
            if is_gap_up and ohlcv[i]['low'] <= prev_close:
                filled_gaps += 1
            elif not is_gap_up and ohlcv[i]['high'] >= prev_close:
                filled_gaps += 1
        if total_gaps > 0:
            self.gap_fill_database[symbol] = {
                'fillRate':   filled_gaps / total_gaps,
                'totalGaps':  total_gaps,
                'filledGaps': filled_gaps,
            }

    # ── Main Scoring Entry Point ──────────────────────────────────────────────

    def compute_gap_score(self, data):
        closes      = data.get('closes', [])
        snapshot    = data.get('snapshot', {})
        highs       = data.get('highs', [])
        lows        = data.get('lows', [])
        volumes     = data.get('volumes', [])
        fundamentals = data.get('fundamentals', {})
        options_data = data.get('optionsData', {})

        _empty = {'gapTier': 0, 'score': 0, 'override': None,
                  'log': [], 'confirmationStrong': False, 'isFadeScenario': False}

        if not closes or len(closes) < 2:
            return _empty

        last_close = closes[-1]
        prev_close = closes[-2]
        today_open = snapshot.get('open') or last_close

        if not today_open or not prev_close:
            return _empty

        gap_percent = (today_open - prev_close) / prev_close * 100
        abs_gap     = abs(gap_percent)
        is_gap_up   = gap_percent > 0

        confirmation_strong = False
        is_fade_scenario    = False

        # ── LAYER 1 R-01: Gap Tier + Multiplier ──────────────────────────────
        if abs_gap < 0.25:
            return {**_empty, 'log': ['Tier 0: No actionable gap']}
        elif abs_gap <= 1:
            gap_tier, multiplier = 1, 0.6
        elif abs_gap <= 3:
            gap_tier, multiplier = 2, 1.0
        elif abs_gap <= 6:
            gap_tier, multiplier = 3, 0.85
        else:
            gap_tier, multiplier = 4, 0.6

        score         = 0
        log           = [f'Gap Tier: {gap_tier} ({gap_percent:.2f}%)']
        override_trigger = None

        adx            = compute_adx(highs, lows, closes) if highs and lows else 20
        rsi            = compute_rsi(closes)
        catalyst_score = snapshot.get('catalyst_score', 0)
        vix            = snapshot.get('india_vix', 15)

        # ── LAYER 6 Early: Hard Overrides ────────────────────────────────────
        if vix > 30:
            return {**_empty, 'gapTier': gap_tier,
                    'override': 'WATCH',
                    'log': log + ['OV-01: India VIX > 30 (Extreme Risk)'],
                    'confirmationStrong': False, 'isFadeScenario': False}

        if snapshot.get('surveillance_tags'):
            return {**_empty, 'gapTier': gap_tier,
                    'override': 'AVOID',
                    'log': log + ['OV-03: Stock under ASM/ESM'],
                    'confirmationStrong': False, 'isFadeScenario': False}

        # ── Gap Type Classification ───────────────────────────────────────────
        has_long_wick = False
        if highs and lows:
            recent_wicks = [
                max(abs(highs[-(5-i)] - closes[-(5-i)]),
                    abs(closes[-(5-i)] - lows[-(5-i)])) / closes[-(5-i)]
                for i in range(min(5, len(highs)))
            ]
            has_long_wick = any(w > 0.03 for w in recent_wicks)

        vwap_val = snapshot.get('vwap') or snapshot.get('avg_price') or last_close
        gap_type = 'Common'
        if catalyst_score >= 6 and adx > 25 and last_close == vwap_val:
            gap_type = 'Breakaway'
            confirmation_strong = True
        elif adx > 30 and last_close != vwap_val:
            gap_type = 'Runaway'
        elif (is_gap_up and rsi > 78) or (not is_gap_up and rsi < 22):
            if has_long_wick:
                gap_type = 'Exhaustion'
                is_fade_scenario = True
        elif adx < 20:
            gap_type = 'Common'
            is_fade_scenario = True
        log.append(f'Gap Type: {gap_type}')

        # ── R-02: Gap Fill Rate ───────────────────────────────────────────────
        symbol      = data.get('symbol', '')
        gap_history = self.gap_fill_database.get(symbol)
        if gap_history and gap_history['totalGaps'] >= 5:
            fr = gap_history['fillRate']
            if gap_type == 'Common' and fr > 0.7:
                score += -8 if is_gap_up else 8
                is_fade_scenario = True
                log.append(f'R-02: High gap fill rate ({fr*100:.0f}%) — fade bias')
            elif gap_type == 'Breakaway' and fr < 0.3:
                score += 10 if is_gap_up else -10
                confirmation_strong = True
                log.append(f'R-02: Low fill rate ({fr*100:.0f}%) — continuation bias')

        # ── R-04: Catalyst Persistence ───────────────────────────────────────
        if catalyst_score >= 6:
            score += 10 if is_gap_up else -10
            confirmation_strong = True
            log.append(f'R-04: Strong Catalyst ({catalyst_score})')

        # ── LAYER 2 R-05: Pre-Open Order Imbalance ───────────────────────────
        pre_buy  = snapshot.get('pre_open_buy_qty')  or snapshot.get('buy_qty')
        pre_sell = snapshot.get('pre_open_sell_qty') or snapshot.get('sell_qty')
        if pre_buy is not None and pre_sell is not None and (pre_buy + pre_sell) > 0:
            imbalance = (pre_buy - pre_sell) / (pre_buy + pre_sell)
            if is_gap_up:
                if imbalance > 0.5:  score += 10
                elif imbalance < -0.3: score -= 10
            else:
                if imbalance < -0.5: score -= 10
                elif imbalance > 0.3: score += 10
            log.append(f'R-05: Pre-Open Imbalance ({imbalance:.2f})')

        # ── R-06: Gift Nifty ─────────────────────────────────────────────────
        gift_prem = snapshot.get('gift_nifty_premium')
        if gift_prem is not None:
            if is_gap_up and gift_prem > 0.5:
                score += 8
                log.append(f'R-06: Gift Nifty +{gift_prem:.2f}% supports gap up')
            elif is_gap_up and gift_prem <= 0:
                score -= 8
                is_fade_scenario = True
            elif not is_gap_up and gift_prem < -0.5:
                score += 8
                log.append(f'R-06: Gift Nifty {gift_prem:.2f}% supports gap down')
            elif not is_gap_up and gift_prem > 0:
                score -= 8
                is_fade_scenario = True

        # ── R-08: VIX Level ──────────────────────────────────────────────────
        if 18 <= vix <= 25:
            if gap_tier < 2 or catalyst_score < 7:
                multiplier *= 0.7
                log.append('R-08: Elevated VIX dampens score')
        elif vix < 13:
            multiplier *= 0.9

        # ── LAYER 3 R-11/R-12: Opening Range ─────────────────────────────────
        or_high = snapshot.get('or_high') or (snapshot.get('ohlc') or {}).get('high', 0)
        or_low  = snapshot.get('or_low')  or (snapshot.get('ohlc') or {}).get('low', 0)
        if or_high and or_low and last_close:
            if is_gap_up:
                if last_close > or_high:
                    score += 15; log.append('R-11: Price > OR_High (bullish continuation)')
                    confirmation_strong = True
                elif last_close < or_low:
                    score -= 15; is_fade_scenario = True; log.append('R-11: Price < OR_Low (gap failure)')
            else:
                if last_close < or_low:
                    score += 15; log.append('R-12: Price < OR_Low (bearish continuation)')
                    confirmation_strong = True
                elif last_close > or_high:
                    score -= 15; is_fade_scenario = True; log.append('R-12: Price > OR_High (V-shape recovery, gap fills)')

        # ── R-13: VWAP Position ───────────────────────────────────────────────
        if vwap_val and vwap_val > 0:
            vwap_diff = (last_close - vwap_val) / vwap_val * 100
            if is_gap_up:
                if last_close > vwap_val:
                    score += 8; log.append(f'R-13: Price > VWAP ({vwap_diff:.2f}% above)')
                elif vwap_diff < -1:
                    score -= 8; is_fade_scenario = True
                    log.append(f'R-13: Price rejected at VWAP ({vwap_diff:.2f}% below)')
            else:
                if last_close < vwap_val:
                    score += 8; log.append(f'R-13: Price < VWAP ({vwap_diff:.2f}% below)')
                elif vwap_diff > 1:
                    score -= 8; is_fade_scenario = True
                    log.append(f'R-13: Price rejected at VWAP ({vwap_diff:.2f}% above)')

        # ── LAYER 4 R-15: Sector Breadth ─────────────────────────────────────
        breadth = snapshot.get('sector_breadth_pct')
        if breadth is not None:
            if is_gap_up:
                if breadth > 70:
                    score += 10; log.append(f'R-15: Sector breadth {breadth}% > 70% (aligned)')
                elif breadth < 40:
                    score -= 10; is_fade_scenario = True
                    log.append(f'R-15: Sector breadth {breadth}% < 40% (divergence)')
            else:
                if breadth < 30:
                    score += 10; log.append(f'R-15: Sector breadth {breadth}% < 30% (aligned)')
                elif breadth > 60:
                    score -= 10; is_fade_scenario = True
                    log.append(f'R-15: Sector breadth {breadth}% > 60% (divergence)')

        # ── R-17: Volume Confirmation ─────────────────────────────────────────
        if volumes and len(volumes) >= 20:
            vol_ratio = compute_volume_ratio(volumes)
            if vol_ratio > 2:
                score += 8; log.append(f'R-17: High volume ({vol_ratio:.1f}x) confirms gap')
            elif vol_ratio < 0.5:
                score -= 8; is_fade_scenario = True
                log.append('R-17: Low volume (weak gap, fade risk)')

        # ── R-18: Holding OR Level ────────────────────────────────────────────
        if or_high and or_low:
            if is_gap_up:
                if last_close > or_high:
                    score += 5; log.append('R-18: Holding above OR_High')
                elif last_close < or_low:
                    score -= 5; log.append('R-18: Failed to hold OR level')
            else:
                if last_close < or_low:
                    score += 5; log.append('R-18: Holding below OR_Low')
                elif last_close > or_high:
                    score -= 5; log.append('R-18: Failed to hold OR level')

        # ── LAYER 5 R-20: Weekly Expiry + Banking ────────────────────────────
        if snapshot.get('is_weekly_expiry') and snapshot.get('sector') == 'Banking':
            multiplier *= 0.7; log.append('R-20: Weekly Expiry + Banking (0.7× multiplier)')

        # ── R-21: Index Divergence ────────────────────────────────────────────
        index_dir = snapshot.get('index_direction')
        if index_dir and index_dir != ('UP' if is_gap_up else 'DOWN'):
            score -= 6; is_fade_scenario = True
            log.append(f'R-21: Index divergence (stock {"up" if is_gap_up else "down"}, index opposite)')

        # ── LAYER 6 Late: Conditional Overrides ──────────────────────────────
        if snapshot.get('promoter_pledge_invoked') and not is_gap_up and score < -10:
            override_trigger = 'STRONG SELL'
            log.append('OV-05: Promoter pledge + gap down + negative momentum')

        earn_surp = snapshot.get('earnings_surprise_pct')
        if earn_surp is not None and earn_surp < -5 and is_gap_up and last_close < vwap_val:
            override_trigger = 'STRONG FADE'; is_fade_scenario = True
            log.append('OV-06: Earnings miss + gap up + VWAP break')

        if gap_tier >= 4 and abs_gap > 5 and highs and lows:
            price_move = ((today_open - lows[-1]) / today_open * 100) if is_gap_up \
                         else ((highs[-1] - today_open) / today_open * 100)
            if price_move > abs_gap * 0.5:
                override_trigger = 'FADE'; is_fade_scenario = True
                log.append(f'OV-07: Extreme gap ({abs_gap:.2f}%) + immediate reversal detected')

        final_score = score * multiplier

        return {
            'gapTier':           gap_tier,
            'score':             final_score,
            'override':          override_trigger,
            'log':               log,
            'gapType':           gap_type,
            'gapFillRate':       gap_history['fillRate'] if gap_history else None,
            'confirmationStrong': confirmation_strong,
            'isFadeScenario':    is_fade_scenario,
        }


# Singleton instance (mirrors JS: globalThis.gapAnalysisEngine = new GapAnalysisEngine())
gap_analysis_engine = GapAnalysisEngine()


# ─────────────────────────────────────────────────────────────────────────────
# Shared helper — used by scoring_engine.py AND entry_validator.py
# Centralises the dynamic gap weighting logic so threshold changes need
# only one edit.
# ─────────────────────────────────────────────────────────────────────────────

def compute_gap_weighted_total(total: float, gap_res: dict) -> tuple:
    """Apply dynamic gap weighting to a score total.

    Args:
        total:   Current score (0-100).
        gap_res: Result dict from gap_analysis_engine.compute_gap_score().

    Returns:
        (new_total, overlay_factor_dict | None)
        overlay_factor_dict is ready to insert into a factors dict.
    """
    if not gap_res or gap_res.get('gapTier', 0) < 1:
        return total, None

    # Dynamic weight — Tier 3+ with strong confirmation → higher gap weight
    w_gap = 0.5 if (gap_res['gapTier'] >= 3 and gap_res.get('confirmationStrong')) else 0.4
    if gap_res.get('isFadeScenario'):
        w_gap = 0.3
    w_core = 1.0 - w_gap

    new_total = total * w_core + gap_res.get('score', 0) * w_gap

    label = (
        f"Gap [{gap_res['override']}]" if gap_res.get('override')
        else f"Gap Tier {gap_res['gapTier']} ({gap_res.get('gapType', '')})"
    )
    overlay = {
        'score':  min(max(gap_res.get('score', 0), -100), 100),
        'max':    100,
        'noData': False,
        'color':  '#D32F2F' if gap_res.get('override') else '#FF9800',
        'label':  label,
    }
    return new_total, overlay

File: app/js/gap-analysis-engine.js
/**
 * TradeSignal — Gap Analysis Rule Engine (Indian Equity Market)
 * Implements refined 6-layer, 24-rule gap analysis institutional overlay.
 * Safer execution with conditional overrides and confirmation requirements.
 * Uses shared TI module for indicator computations.
 */

class GapAnalysisEngine {
  constructor() {
    // Gap fill history — populated from historical OHLCV via buildGapFillHistory()
    // Format: { symbol: { fillRate: 0-1, totalGaps: N, filledGaps: N } }
    this.gapFillDatabase = {};
    
    // Track immediate reversal detection (first 10 min)
    this.reversalDetector = {};
    
    // Auto-initialize gap fill history on startup
    this.initializationFlag = false;
  }

  /**
   * Initialize gap fill history from market snapshot data.
   * Call this once after initial snapshot/OHLCV load.
   */
  initializeGapFillHistory(symbol, ohlcv) {
    if (!symbol || this.gapFillDatabase[symbol]) return;  // Already initialized or invalid
    this.buildGapFillHistory(symbol, ohlcv);
    this.initializationFlag = true;
  }

  /**
   * Build gap fill history from historical OHLCV data.
   * A "gap" is an open price deviating > 0.5% from previous close.
   * A filled gap means intraday price returned to the previous close level.
   * Call this once per stock after loading OHLCV data.
   */
  buildGapFillHistory(symbol, ohlcv) {
    if (!ohlcv || ohlcv.length < 10) return;
    let totalGaps = 0, filledGaps = 0;
    for (let i = 1; i < ohlcv.length; i++) {
      const prevClose = ohlcv[i - 1].close;
      const open = ohlcv[i].open;
      const gapPct = Math.abs((open - prevClose) / prevClose * 100);
      if (gapPct < 0.5) continue;   // skip insignificant gaps
      totalGaps++;
      const isGapUp = open > prevClose;
      // Check if gap was filled intraday
      if (isGapUp && ohlcv[i].low <= prevClose) filledGaps++;
      else if (!isGapUp && ohlcv[i].high >= prevClose) filledGaps++;
    }
    if (totalGaps > 0) {
      this.gapFillDatabase[symbol] = {
        fillRate: filledGaps / totalGaps,
        totalGaps,
        filledGaps
      };
    }
  }

  computeGapScore(data) {
    const { closes, snapshot = {}, highs, lows, volumes, fundamentals = {}, optionsData = {} } = data;
    
    // Validate inputs
    if (!closes || closes.length < 2) {
      return { gapTier: 0, score: 0, override: null, log: [], confirmationStrong: false, isFadeScenario: false };
    }

    const lastClose = closes[closes.length - 1];
    const prevClose = closes[closes.length - 2];
    const todayOpen = snapshot.open || lastClose;
    
    if (!todayOpen || !prevClose) {
      return { gapTier: 0, score: 0, override: null, log: [], confirmationStrong: false, isFadeScenario: false };
    }

    const gapPercent = ((todayOpen - prevClose) / prevClose) * 100;
    const absGap = Math.abs(gapPercent);
    const isGapUp = gapPercent > 0;
    
    let gapTier = 0;
    let multiplier = 1.0;
    let confirmationStrong = false;
    let isFadeScenario = false;
    
    // ── LAYER 1 (R-01): Gap Size Tiering (Refined Multipliers) ──
    if (absGap < 0.25) {
      return { gapTier: 0, score: 0, override: null, log: ['Tier 0: No actionable gap'], confirmationStrong: false, isFadeScenario: false };
    } else if (absGap <= 1) {
      gapTier = 1; 
      multiplier = 0.6;  // REFINED: was 0.5
    } else if (absGap <= 3) {
      gapTier = 2; 
      multiplier = 1.0;
    } else if (absGap <= 6) {
      gapTier = 3; 
      multiplier = 0.85;  // REFINED: was 0.7
    } else {
      gapTier = 4; 
      multiplier = 0.6;
    }

    let score = 0;
    let log = [`Gap Tier: ${gapTier} (${gapPercent.toFixed(2)}%)`];
    let overrideTrigger = null;

    // ── Compute technical indicators early ──
    const adx = TI.computeADX(highs, lows, closes);
    const rsi = TI.computeRSI(closes);
    const catalystScore = snapshot.catalyst_score || 0;
    const vix = snapshot.india_vix || 15;

    // ── LAYER 6 (Early Checks): Hard Overrides (Conditional) ──
    
    // OV-01: India VIX > 30 → No fresh trades
    if (vix > 30) {
      return { gapTier, score: 0, override: 'WATCH', log: [...log, 'OV-01: India VIX > 30 (Extreme Risk)'], confirmationStrong: false, isFadeScenario: false };
    }
    
    // OV-03: Stock under ASM/ESM → Mandatory avoid
    if (snapshot.surveillance_tags && snapshot.surveillance_tags.length > 0) {
      return { gapTier, score: 0, override: 'AVOID', log: [...log, 'OV-03: Stock under ASM/ESM'], confirmationStrong: false, isFadeScenario: false };
    }

    // ── LAYER 1 (cont): Gap Type Classification (Refined) ──
    let gapType = 'Common';
    let hasLongWick = false;
    
    // Check for long wick (high/low significantly away from close)
    if (highs.length > 0 && lows.length > 0) {
      const recent_wicks = highs.slice(-5).map((h, i) => {
        const l = lows[lows.length - 5 + i];
        const c = closes[closes.length - 5 + i];
        return Math.max(Math.abs(h - c), Math.abs(c - l)) / c;
      });
      hasLongWick = recent_wicks.some(w => w > 0.03);  // >3% wick
    }
    
    // Refined Gap Type Classification
    if (catalystScore >= 6 && adx > 25 && lastClose === (snapshot.vwap || lastClose)) {
      gapType = 'Breakaway';   // Catalyst + Trend + VWAP held
      confirmationStrong = true;
    } else if (adx > 30 && lastClose !== (snapshot.vwap || lastClose)) {
      gapType = 'Runaway';      // Strong trend without VWAP retest
    } else if ((isGapUp && rsi > 78) || (!isGapUp && rsi < 22)) {
      if (hasLongWick) {
        gapType = 'Exhaustion';  // Overbought/sold with long wick
        isFadeScenario = true;
      }
    } else if (adx < 20) {
      gapType = 'Common';
      isFadeScenario = true;    // Low ADX = high gap fill probability
    }
    log.push(`Gap Type: ${gapType}`);

    // ── LAYER 1 (cont): Gap Fill Rate (R-02) ──
    const symbol = data.symbol || '';
    const gapHistory = this.gapFillDatabase[symbol];
    if (gapHistory && gapHistory.totalGaps >= 5) {
      const fr = gapHistory.fillRate;
      if (gapType === 'Common' && fr > 0.7) {
        // High fill rate on common gaps → fade bias
        score += isGapUp ? -8 : 8;
        isFadeScenario = true;
        log.push(`R-02: High gap fill rate (${(fr * 100).toFixed(0)}%) — fade bias`);
      } else if (gapType === 'Breakaway' && fr < 0.3) {
        // Low fill rate on breakaway → continuation bias
        score += isGapUp ? 10 : -10;
        confirmationStrong = true;
        log.push(`R-02: Low fill rate (${(fr * 100).toFixed(0)}%) — continuation bias`);
      }
    }

    // ── LAYER 1 (cont): Catalyst Persistence (R-04) ──
    if (catalystScore >= 6) {
      score += isGapUp ? 10 : -10;
      confirmationStrong = true;
      log.push(`R-04: Strong Catalyst (${catalystScore})`);
    }

    // ── LAYER 2: Pre-Open Signals (R-05 to R-09) ──
    const preOpenBuy = snapshot.pre_open_buy_qty ?? snapshot.buy_qty;
    const preOpenSell = snapshot.pre_open_sell_qty ?? snapshot.sell_qty;
    if (preOpenBuy != null && preOpenSell != null && (preOpenBuy + preOpenSell) > 0) {
      const bQty = preOpenBuy;
      const sQty = preOpenSell;
      const imbalance = (bQty - sQty) / (bQty + sQty || 1);
      
      // Capped to ±10 (refined: was ±12)
      if (isGapUp) {
        if (imbalance > 0.5) score += 10;
        else if (imbalance < -0.3) score -= 10;
      } else {
        if (imbalance < -0.5) score -= 10;
        else if (imbalance > 0.3) score += 10;
      }
      log.push(`R-05: Pre-Open Imbalance (${imbalance.toFixed(2)})`);
    }

    // ── R-06: Gift Nifty Premium/Discount (aligned check) ──
    if (snapshot.gift_nifty_premium !== undefined) {
      const giftPrem = snapshot.gift_nifty_premium;
      if (isGapUp && giftPrem > 0.5) {
        score += 8;  // Gift aligned with gap up
        log.push(`R-06: Gift Nifty +${giftPrem.toFixed(2)}% supports gap up`);
      } else if (isGapUp && giftPrem <= 0) {
        score -= 8;  // Divergence: gap up but Gift down
        isFadeScenario = true;
      } else if (!isGapUp && giftPrem < -0.5) {
        score += 8;  // Gift aligned with gap down
        log.push(`R-06: Gift Nifty ${giftPrem.toFixed(2)}% supports gap down`);
      } else if (!isGapUp && giftPrem > 0) {
        score -= 8;  // Divergence: gap down but Gift up
        isFadeScenario = true;
      }
    }

    // ── R-08: VIX Level (Refined multiplier: 0.7 instead of 0.5) ──
    if (vix >= 18 && vix <= 25) {
      if (gapTier < 2 || catalystScore < 7) {
        multiplier *= 0.7;  // REFINED: was 0.5
        log.push('R-08: Elevated VIX dampens score');
      }
    } else if (vix < 13) {
      multiplier *= 0.9;
    }

    // ── LAYER 3: Opening Range & VWAP Validation (Apply after 5–10 min) ──
    const orHigh = snapshot.or_high || (snapshot.ohlc ? snapshot.ohlc.high : 0);
    const orLow = snapshot.or_low || (snapshot.ohlc ? snapshot.ohlc.low : 0);
    
    if (orHigh && orLow && lastClose) {
      if (isGapUp) {
        // R-11 (Gap Up): Bullish continuation
        if (lastClose > orHigh) {
          score += 15;  // Price breaks above OR_High
          log.push('R-11: Price > OR_High (bullish continuation)');
          confirmationStrong = true;
        } else if (lastClose < orLow) {
          score -= 15;  // Gap failure → bearish
          isFadeScenario = true;
          log.push('R-11: Price < OR_Low (gap failure)');
        }
      } else {
        // R-12 (Gap Down): Refined logic
        if (lastClose < orLow) {
          score += 15;  // REFINED: was -15 (bearish continuation now POSITIVE for gap down)
          log.push('R-12: Price < OR_Low (bearish continuation)');
          confirmationStrong = true;
        } else if (lastClose > orHigh) {
          score -= 15;  // REFINED: was +20 (V-shape recovery now NEGATIVE)
          isFadeScenario = true;
          log.push('R-12: Price > OR_High (V-shape recovery, gap fills)');
        }
      }
    }

    // ── R-13: VWAP Position (Refined with rejection logic) ──
    const vwap = snapshot.vwap || snapshot.avg_price;
    if (vwap && vwap > 0) {
      const vwapDiff = ((lastClose - vwap) / vwap) * 100;
      
      if (isGapUp) {
        if (lastClose > vwap) {
          score += 8;    // Above VWAP: strength
          log.push(`R-13: Price > VWAP (${vwapDiff.toFixed(2)}% above)`);
        } else if (vwapDiff > -1) {
          score -= 0;    // Near VWAP: neutral
        } else {
          score -= 8;    // VWAP rejection (below)
          isFadeScenario = true;
          log.push(`R-13: Price rejected at VWAP (${vwapDiff.toFixed(2)}% below)`);
        }
      } else {
        if (lastClose < vwap) {
          score += 8;    // Below VWAP: weakness confirmation
          log.push(`R-13: Price < VWAP (${vwapDiff.toFixed(2)}% below)`);
        } else if (vwapDiff < 1) {
          score -= 0;    // Near VWAP: neutral
        } else {
          score -= 8;    // VWAP rejection (above)
          isFadeScenario = true;
          log.push(`R-13: Price rejected at VWAP (${vwapDiff.toFixed(2)}% above)`);
        }
      }
    }

    // ── LAYER 4: Intraday Continuation (R-15 to R-18) ──
    
    // R-15: Sector Breadth
    if (snapshot.sector_breadth_pct !== undefined) {
      if (isGapUp) {
        if (snapshot.sector_breadth_pct > 70) {
          score += 10;
          log.push(`R-15: Sector breadth ${snapshot.sector_breadth_pct}% > 70% (aligned)`);
        } else if (snapshot.sector_breadth_pct < 40) {
          score -= 10;
          isFadeScenario = true;
          log.push(`R-15: Sector breadth ${snapshot.sector_breadth_pct}% < 40% (divergence)`);
        }
      } else {
        if (snapshot.sector_breadth_pct < 30) {
          score += 10;
          log.push(`R-15: Sector breadth ${snapshot.sector_breadth_pct}% < 30% (aligned)`);
        } else if (snapshot.sector_breadth_pct > 60) {
          score -= 10;
          isFadeScenario = true;
          log.push(`R-15: Sector breadth ${snapshot.sector_breadth_pct}% > 60% (divergence)`);
        }
      }
    }

    // R-17: Volume Confirmation (Refined: low volume now -8, was -5)
    if (volumes && volumes.length >= 20) {
      const volRatio = TI.computeVolumeRatio(volumes);
      if (volRatio > 2) {
        score += 8;
        log.push(`R-17: High volume (${volRatio.toFixed(1)}x) confirms gap`);
      } else if (volRatio < 0.5) {
        score -= 8;   // REFINED: was -5
        isFadeScenario = true;
        log.push('R-17: Low volume (weak gap, fade risk)');
      }
    }

    // R-18: Price Holding OR Level
    if (orHigh && orLow) {
      if (isGapUp) {
        if (lastClose > orHigh) {
          score += 5;
          log.push('R-18: Holding above OR_High');
        } else if (lastClose < orLow) {
          score -= 5;
          log.push('R-18: Failed to hold OR level');
        }
      } else {
        if (lastClose < orLow) {
          score += 5;
          log.push('R-18: Holding below OR_Low');
        } else if (lastClose > orHigh) {
          score -= 5;
          log.push('R-18: Failed to hold OR level');
        }
      }
    }

    // ── LAYER 5: India Specifics (R-20 to R-21) ──
    
    // R-20: Weekly Expiry + Banking Sector (Refined: 0.7 instead of 0.65)
    if (snapshot.is_weekly_expiry && snapshot.sector === 'Banking') {
      multiplier *= 0.7;  // REFINED: was 0.65
      log.push('R-20: Weekly Expiry + Banking (0.7× multiplier)');
    }

    // R-21: Index Divergence (NEW)
    if (snapshot.index_direction && snapshot.index_direction !== (isGapUp ? 'UP' : 'DOWN')) {
      score -= 6;
      isFadeScenario = true;
      log.push(`R-21: Index divergence (stock ${isGapUp ? 'up' : 'down'}, index opposite)`);
    }

    // ── LAYER 6 (cont): Conditional Overrides with Confirmation ──
    
    // OV-05: Promoter Pledge Invocation + Gap Down + Continuation (Refined)
    if (snapshot.promoter_pledge_invoked && !isGapUp && score < -10) {
      overrideTrigger = 'STRONG SELL';
      log.push('OV-05: Promoter pledge + gap down + negative momentum');
    }

    // OV-06: Earnings Miss + Gap Up + VWAP Break (Refined with validation)
    if (snapshot.earnings_surprise_pct !== undefined) {
      if (snapshot.earnings_surprise_pct < -5 && isGapUp && lastClose < vwap) {
        overrideTrigger = 'STRONG FADE';
        isFadeScenario = true;
        log.push('OV-06: Earnings miss + gap up + VWAP break');
      }
    }

    // OV-07: Immediate Reversal (Gap >5% + First 10 min reversal) — NEW
    if (gapTier >= 4 && absGap > 5) {
      // Check if there's immediate reversal signal (high within first bar if gap up, low if gap down)
      const priceMoveWithinGap = isGapUp 
        ? ((todayOpen - lows[lows.length - 1]) / todayOpen * 100)
        : ((highs[highs.length - 1] - todayOpen) / todayOpen * 100);
      
      if (priceMoveWithinGap > absGap * 0.5) {
        overrideTrigger = 'FADE';
        isFadeScenario = true;
        log.push(`OV-07: Extreme gap (${absGap.toFixed(2)}%) + immediate reversal detected`);
      }
    }

    // ── Final Score Calculation ──
    const finalScore = score * multiplier;

    return { 
      gapTier, 
      score: finalScore, 
      override: overrideTrigger, 
      log, 
      gapType, 
      gapFillRate: gapHistory?.fillRate,
      confirmationStrong,     // NEW: high confidence signal
      isFadeScenario          // NEW: indicates fade/reversal probability
    };
  }
}

globalThis.gapAnalysisEngine = new GapAnalysisEngine();
